satpy/dataset.py

```diff
@@ -149,163 +358,315 @@ def _share_metadata_key_list_arrays(values):
+        new_id_keys = id_keys.copy()
+        for key, val in id_keys.items():
+            if not val:
+                continue
+            if 'enum' in val and 'type' in val:
+                raise ValueError('Cannot have both type and enum for the same id key.')
+            new_val = copy(val)
+            if 'enum' in val:
+                new_val['type'] = ValueList(key, ' '.join(new_val.pop('enum')))
+            new_id_keys[key] = new_val
+        return new_id_keys
+
+    def convert_dict(self, keyvals):
+        """Convert a dictionary's values to the types defined in this object's id_keys."""
+        curated = {}
+        if not keyvals:
+            return curated
+        for key, val in self._id_keys.items():
+            if val is not None:
+                if key in keyvals or val.get('default') or val.get('required'):
+                    curated_val = keyvals.get(key, val.get('default'))
+                    if 'required' in val and curated_val is None:
+                        raise ValueError('Required field {} missing.'.format(key))
+                    if 'type' in val:
+                        curated[key] = val['type'].convert(curated_val)
+                    elif curated_val is not None:
+                        curated[key] = curated_val
+            else:
+                try:
+                    curated_val = keyvals[key]
+                except KeyError:
+                    pass
+                else:
+                    if curated_val is not None:
+                        curated[key] = curated_val
+        return curated
+
+    def from_dict(self, keyvals):
+        """Create a DataID from a dictionary."""
+        return self.__class__(self._id_keys, **keyvals)
 
-    def __new__(cls, *args, **kwargs):
-        """Create new DatasetID."""
-        ret = super(DatasetID, cls).__new__(cls, *args, **kwargs)
-        if ret.modifiers is not None and not isinstance(ret.modifiers, tuple):
-            raise TypeError("'DatasetID' modifiers must be a tuple or None, "
-                            "not {}".format(type(ret.modifiers)))
-        return ret
+    @classmethod
+    def from_dataarray(cls, array, default_keys=minimal_default_keys_config):
+        """Create a DataID from a dataarray."""
+        try:
+            id_keys = array.attrs['_satpy_id'].id_keys
+        except KeyError:
+            id_keys = array.attrs.get('_satpy_id_keys', default_keys)
+        return cls(id_keys, **array.attrs)
+
+    def _asdict(self):
+        return dict(self.items())
+
+    def to_dict(self):
+        """Convert the ID to a dict."""
+        return self._asdict()
 
-    @staticmethod
-    def name_match(a, b):
-        """Return if two string names are equal.
+    def __getattr__(self, key):
+        """Support old syntax for getting items."""
+        if key != '_id_keys' and key in self._id_keys:
+            warnings.warn('Attribute access to DataIDs is deprecated, use key access instead.')
```

Comment:
Please add a `category=DeprecationWarning` so that users can silence `DeprecationWarning`s such as this one specifically without silencing other warnings.

(I thought I had noted this already but can't find it now)

Reply:
The problem with DeprecationWarning is that it isn't shown to the user by default (iirc). The default, UserWarning, is and seems to become semi-standard in scientific python packages (ex. dask and xarray) for deprecation warnings.

Reply:
`DeprecationWarning` is shown by default since Python 3.7, `PendingDeprecationWarning` is not.  

If community packages are showing `UserWarning` for deprecations instead of `DeprecationWarning` because the latter was not shown by default in versions 3.2–3.6 (inclusive) then that defeats the point of a `DeprecationWarning` category to exist.  Although I can see the point in Satpy following the example of dask and xarray, I rather hope that `DeprecationWarning`s return to the warnings shown by default will trigger other packages to use it again.

Reply:
Meh, `DeprecationWarning` is indeed only shown by default only when triggered in `__main__`.  If for that reason packages are avoiding to use it then that defeats the point of this warning category to exist in the first place :-/

Reply:
I think the main issue is that UserWarning should probably be used for *users* that are using a deprecated feature. DeprecationWarning should be for developers that are using deprecated features/interfaces and should be developing with the proper flags/settings to show DeprecationWarnings. This is very confusing when the users are developers and are using a library.

Reply:
FWIW, Satpy currently uses `DeprecationWarning` in 7 places, `pyresample` in 17 places.

Reply:
Other point: consider adding a `stacklevel=2` argument.  That way, instead of getting a warning message such as

`/data/gholl/miniconda3/envs/py38/lib/python3.8/site-packages/satpy/dataset.py:496: UserWarning: Attribute access to DataIDs is deprecated, use key access instead.`

the user should get a warning message with points to the line using the attribute access (I think `stacklevel=2` would do that).